gobblin-core/src/main/java/gobblin/writer/PartitionedDataWriter.java

```diff
@@ -0,0 +1,132 @@
+    });
+
+    if(state.contains(ConfigurationKeys.WRITER_PARTITIONER_CLASS)) {
+      Preconditions.checkArgument(builder instanceof PartitionAwareDataWriterBuilder,
+          String.format("%s was specified but the writer %s does not support partitioning.",
+              ConfigurationKeys.WRITER_PARTITIONER_CLASS, builder.getClass().getCanonicalName()));
+
+      try {
+        this.shouldPartition = true;
+        this.builder = Optional.of(PartitionAwareDataWriterBuilder.class.cast(builder));
+        this.partitioner = Optional.of(
+            WriterPartitioner.class.cast(ConstructorUtils.invokeConstructor(
+                    Class.forName(state.getProp(ConfigurationKeys.WRITER_PARTITIONER_CLASS)), state)));
+        this.builder.get().validatePartitionSchema(this.partitioner.get().partitionSchema());
+      } catch(ReflectiveOperationException roe) {
+        throw new IOException(roe);
+      }
+    } else {
+      this.shouldPartition = false;
+      this.partitionWriters.put(NON_PARTITIONED_WRITER_KEY,
+          new InstrumentedDataWriterDecorator<D>(builder.build(), state));
+      this.partitioner = Optional.absent();
+      this.builder = Optional.absent();
+    }
+  }
+
+  @Override public void write(D record) throws IOException {
+    try {
+      GenericRecord partition = this.shouldPartition ? this.partitioner.get().partitionForRecord(record) :
+          NON_PARTITIONED_WRITER_KEY;
+      DataWriter<D> writer = this.partitionWriters.get(partition);
+      writer.write(record);
+    } catch(ExecutionException ee) {
+      throw new IOException(ee);
+    }
+  }
+
+  @Override public void commit() throws IOException {
+    for(Map.Entry<GenericRecord, DataWriter<D>> entry : this.partitionWriters.asMap().entrySet()) {
+      entry.getValue().commit();
+    }
+  }
+
+  @Override public void cleanup() throws IOException {
+    for(Map.Entry<GenericRecord, DataWriter<D>> entry : this.partitionWriters.asMap().entrySet()) {
+      entry.getValue().cleanup();
+    }
+  }
+
+  @Override public long recordsWritten() {
+    long totalRecords = 0;
+    for(Map.Entry<GenericRecord, DataWriter<D>> entry : this.partitionWriters.asMap().entrySet()) {
+      totalRecords += entry.getValue().recordsWritten();
+    }
+    return totalRecords;
+  }
+
+  @Override public long bytesWritten() throws IOException {
+    long totalBytes = 0;
+    for(Map.Entry<GenericRecord, DataWriter<D>> entry : this.partitionWriters.asMap().entrySet()) {
+      totalBytes += entry.getValue().bytesWritten();
+    }
+    return totalBytes;
+  }
+
+  @Override public void close() throws IOException {
+    for(Map.Entry<GenericRecord, DataWriter<D>> entry : this.partitionWriters.asMap().entrySet()) {
+      entry.getValue().close();
+    }
+  }
+
+  private DataWriter<D> createPartitionWriter(GenericRecord partition) throws IOException {
+    if(!this.builder.isPresent()) {
+      throw new IOException("Writer builder not found. This is an error in the code.");
```

Comment:
The partition is irrelevant. The only way this condition can be reached is if the PartitionedWriter was constructed incorrectly, and in that case this will happen in the first partition, regardless of the specific value of the partition.